diff --git a/third_party/php8.4-src/ext/libxml/libxml.c b/third_party/php8.4-src/ext/libxml/libxml.c
index f07b321..4a1a2da 100644
--- a/third_party/php8.4-src/ext/libxml/libxml.c
+++ b/third_party/php8.4-src/ext/libxml/libxml.c
@@ -48,6 +48,17 @@
 
 #include "libxml_arginfo.h"
 
+#ifdef LIBXML_DYNAMIC_LOAD
+__attribute__((weak)) const char *xmlParserVersion = {0};
+__attribute__((weak)) void (*xmlFree)(void *);
+__attribute__((weak)) void *(*xmlMalloc)(size_t);
+__attribute__((weak)) void *(*xmlMallocAtomic)(size_t);
+__attribute__((weak)) char *(*xmlMemStrdup)(const char *);
+__attribute__((weak)) void *(*xmlRealloc)(void *, size_t);
+__attribute__((weak)) void (*xmlGenericError)(void *, const char *, ...);
+__attribute__((weak)) void (*xmlStructuredError)(void *, xmlErrorPtr);
+#endif
+
 /* a true global for initialization */
 static int _php_libxml_initialized = 0;
 static int _php_libxml_per_request_initialization = 1;
@@ -938,6 +949,29 @@ PHP_LIBXML_API void php_libxml_switch_context(zval *context, zval *oldcontext)
 
 static PHP_MINIT_FUNCTION(libxml)
 {
+	if (xmlParserVersion == NULL) {
+		void *handle = DL_LOAD("libxml2.so");
+		const char** xmlParserVersionPtr = (const char**)dlsym(handle, "xmlParserVersion");
+		if (xmlParserVersionPtr) {
+			xmlParserVersion = *xmlParserVersionPtr;
+		}
+	}
+
+	if (xmlParserVersion == NULL) {
+		zend_disable_class("LibXMLError", sizeof("LibXMLError") - 1);
+		zend_disable_functions(
+			"libxml_set_streams_context,"
+			"libxml_use_internal_errors,"
+			"libxml_get_last_error,"
+			"libxml_get_errors,"
+			"libxml_clear_errors,"
+			"libxml_disable_entity_loader,"
+			"libxml_set_external_entity_loader,"
+			"libxml_get_external_entity_loader,"
+		);
+		return SUCCESS;
+	}
+
 	php_libxml_initialize();
 
 	register_libxml_symbols(module_number);
@@ -973,6 +1007,8 @@ static PHP_MINIT_FUNCTION(libxml)
 
 static PHP_RINIT_FUNCTION(libxml)
 {
+	if (xmlParserVersion == NULL) return SUCCESS;
+
 	if (_php_libxml_per_request_initialization) {
 		/* report errors via handler rather than stderr */
 		xmlSetGenericErrorFunc(NULL, php_libxml_error_handler);
@@ -1000,6 +1036,8 @@ static PHP_RSHUTDOWN_FUNCTION(libxml)
 
 static PHP_MSHUTDOWN_FUNCTION(libxml)
 {
+	if (xmlParserVersion == NULL) return SUCCESS;
+
 	if (!_php_libxml_per_request_initialization) {
 		xmlSetGenericErrorFunc(NULL, NULL);
 
@@ -1013,6 +1051,8 @@ static PHP_MSHUTDOWN_FUNCTION(libxml)
 
 static zend_result php_libxml_post_deactivate(void)
 {
+	if (xmlParserVersion == NULL) return SUCCESS;
+
 	/* reset libxml generic error handling */
 	if (_php_libxml_per_request_initialization) {
 		xmlSetGenericErrorFunc(NULL, NULL);
@@ -1038,6 +1078,13 @@ static zend_result php_libxml_post_deactivate(void)
 
 static PHP_MINFO_FUNCTION(libxml)
 {
+	if (xmlParserVersion == NULL) {
+		php_info_print_table_start();
+		php_info_print_table_row(2, "libXML support", "NOT ACTIVE");
+		php_info_print_table_end();
+		return;
+	}
+
 	php_info_print_table_start();
 	php_info_print_table_row(2, "libXML support", "active");
 	php_info_print_table_row(2, "libXML Compiled Version", LIBXML_DOTTED_VERSION);
@@ -1050,6 +1097,8 @@ static PHP_MINFO_FUNCTION(libxml)
 /* {{{ Set the streams context for the next libxml document load or write */
 PHP_FUNCTION(libxml_set_streams_context)
 {
+	if (xmlParserVersion == NULL) return;
+
 	zval *arg;
 
 	ZEND_PARSE_PARAMETERS_START(1, 1)
@@ -1073,6 +1122,8 @@ PHP_LIBXML_API bool php_libxml_uses_internal_errors(void)
 /* {{{ Disable libxml errors and allow user to fetch error information as needed */
 PHP_FUNCTION(libxml_use_internal_errors)
 {
+	if (xmlParserVersion == NULL) return;
+
 	bool use_errors, use_errors_is_null = true;
 
 	ZEND_PARSE_PARAMETERS_START(0, 1)
@@ -1126,6 +1177,8 @@ static void php_libxml_create_error_object(zval *return_value, const xmlError *e
 /* {{{ Retrieve last error from libxml */
 PHP_FUNCTION(libxml_get_last_error)
 {
+	if (xmlParserVersion == NULL) return;
+
 	ZEND_PARSE_PARAMETERS_NONE();
 
 	const xmlError *error;
@@ -1147,6 +1200,8 @@ PHP_FUNCTION(libxml_get_last_error)
 /* {{{ Retrieve array of errors */
 PHP_FUNCTION(libxml_get_errors)
 {
+	if (xmlParserVersion == NULL) return;
+
 	xmlErrorPtr error;
 
 	ZEND_PARSE_PARAMETERS_NONE();
@@ -1170,6 +1225,8 @@ PHP_FUNCTION(libxml_get_errors)
 /* {{{ Clear last error from libxml */
 PHP_FUNCTION(libxml_clear_errors)
 {
+	if (xmlParserVersion == NULL) return;
+
 	ZEND_PARSE_PARAMETERS_NONE();
 
 	xmlResetLastError();
@@ -1190,6 +1247,8 @@ PHP_LIBXML_API bool php_libxml_disable_entity_loader(bool disable) /* {{{ */
 /* {{{ Disable/Enable ability to load external entities */
 PHP_FUNCTION(libxml_disable_entity_loader)
 {
+	if (xmlParserVersion == NULL) return;
+
 	bool disable = 1;
 
 	ZEND_PARSE_PARAMETERS_START(0, 1)
@@ -1204,6 +1263,8 @@ PHP_FUNCTION(libxml_disable_entity_loader)
 /* {{{ Changes the default external entity loader */
 PHP_FUNCTION(libxml_set_external_entity_loader)
 {
+	if (xmlParserVersion == NULL) return;
+
 	zend_fcall_info			fci;
 	zend_fcall_info_cache	fcc;
 
@@ -1225,6 +1286,8 @@ PHP_FUNCTION(libxml_set_external_entity_loader)
 /* {{{ Get the current external entity loader, or null if the default loader is installer. */
 PHP_FUNCTION(libxml_get_external_entity_loader)
 {
+	if (xmlParserVersion == NULL) return;
+
 	ZEND_PARSE_PARAMETERS_NONE();
 
 	if (ZEND_FCC_INITIALIZED(LIBXML(entity_loader_callback))) {
diff --git a/third_party/php8.4-src/ext/openssl/openssl.c b/third_party/php8.4-src/ext/openssl/openssl.c
index be68216..a8ba810 100644
--- a/third_party/php8.4-src/ext/openssl/openssl.c
+++ b/third_party/php8.4-src/ext/openssl/openssl.c
@@ -1336,8 +1336,8 @@ PHP_MINIT_FUNCTION(openssl)
 	/* override the default tcp socket provider */
 	php_stream_xport_register("tcp", php_openssl_ssl_socket_factory);
 
-	php_register_url_stream_wrapper("https", &php_stream_http_wrapper);
-	php_register_url_stream_wrapper("ftps", &php_stream_ftp_wrapper);
+	// php_register_url_stream_wrapper("https", &php_stream_http_wrapper);
+	// php_register_url_stream_wrapper("ftps", &php_stream_ftp_wrapper);
 
 	REGISTER_INI_ENTRIES();
 
diff --git a/third_party/php8.4-src/ext/pdo_sqlite/pdo_sqlite.c b/third_party/php8.4-src/ext/pdo_sqlite/pdo_sqlite.c
index bc47c15..e39e120 100644
--- a/third_party/php8.4-src/ext/pdo_sqlite/pdo_sqlite.c
+++ b/third_party/php8.4-src/ext/pdo_sqlite/pdo_sqlite.c
@@ -22,8 +22,8 @@
 #include "php_ini.h"
 #include "ext/standard/info.h"
 #include "SAPI.h"
-#include "ext/pdo/php_pdo.h"
-#include "ext/pdo/php_pdo_driver.h"
+#include "../ext/pdo/php_pdo.h"
+#include "../ext/pdo/php_pdo_driver.h"
 #include "php_pdo_sqlite.h"
 #include "php_pdo_sqlite_int.h"
 #include "zend_exceptions.h"
diff --git a/third_party/php8.4-src/ext/pdo_sqlite/sqlite_driver.c b/third_party/php8.4-src/ext/pdo_sqlite/sqlite_driver.c
index 3c2b4ef..818d913 100644
--- a/third_party/php8.4-src/ext/pdo_sqlite/sqlite_driver.c
+++ b/third_party/php8.4-src/ext/pdo_sqlite/sqlite_driver.c
@@ -21,8 +21,8 @@
 #include "php.h"
 #include "php_ini.h"
 #include "ext/standard/info.h"
-#include "ext/pdo/php_pdo.h"
-#include "ext/pdo/php_pdo_driver.h"
+#include "../ext/pdo/php_pdo.h"
+#include "../ext/pdo/php_pdo_driver.h"
 #include "php_pdo_sqlite.h"
 #include "php_pdo_sqlite_int.h"
 #include "zend_exceptions.h"
diff --git a/third_party/php8.4-src/ext/pdo_sqlite/sqlite_statement.c b/third_party/php8.4-src/ext/pdo_sqlite/sqlite_statement.c
index 29309ac..728b862 100644
--- a/third_party/php8.4-src/ext/pdo_sqlite/sqlite_statement.c
+++ b/third_party/php8.4-src/ext/pdo_sqlite/sqlite_statement.c
@@ -21,8 +21,8 @@
 #include "php.h"
 #include "php_ini.h"
 #include "ext/standard/info.h"
-#include "ext/pdo/php_pdo.h"
-#include "ext/pdo/php_pdo_driver.h"
+#include "../ext/pdo/php_pdo.h"
+#include "../ext/pdo/php_pdo_driver.h"
 #include "php_pdo_sqlite.h"
 #include "php_pdo_sqlite_int.h"
 
diff --git a/third_party/php8.4-src/ext/standard/basic_functions.c b/third_party/php8.4-src/ext/standard/basic_functions.c
index dbbf989..c947419 100644
--- a/third_party/php8.4-src/ext/standard/basic_functions.c
+++ b/third_party/php8.4-src/ext/standard/basic_functions.c
@@ -357,8 +357,8 @@ PHP_MINIT_FUNCTION(basic) /* {{{ */
 	php_register_url_stream_wrapper("glob", &php_glob_stream_wrapper);
 #endif
 	php_register_url_stream_wrapper("data", &php_stream_rfc2397_wrapper);
-	php_register_url_stream_wrapper("http", &php_stream_http_wrapper);
-	php_register_url_stream_wrapper("ftp", &php_stream_ftp_wrapper);
+	// php_register_url_stream_wrapper("http", &php_stream_http_wrapper);
+	// php_register_url_stream_wrapper("ftp", &php_stream_ftp_wrapper);
 
 	return SUCCESS;
 }
diff --git a/third_party/php8.4-src/ext/standard/credits.c b/third_party/php8.4-src/ext/standard/credits.c
index 4e8722d..1343330 100644
--- a/third_party/php8.4-src/ext/standard/credits.c
+++ b/third_party/php8.4-src/ext/standard/credits.c
@@ -33,6 +33,11 @@ PHPAPI ZEND_COLD void php_print_credits(int flag) /* {{{ */
 		PUTS("PHP Credits\n");
 	}
 
+	php_info_print_table_start();
+	php_info_print_table_header(1, "php-wasm");
+	php_info_print_table_row(1, "Sean Morris & Oraoto");
+	php_info_print_table_end();
+
 	if (flag & PHP_CREDITS_GROUP) {
 		/* Group */
 
diff --git a/third_party/php8.4-src/ext/standard/credits_ext.h b/third_party/php8.4-src/ext/standard/credits_ext.h
index 49b263c..d8e38a8 100644
--- a/third_party/php8.4-src/ext/standard/credits_ext.h
+++ b/third_party/php8.4-src/ext/standard/credits_ext.h
@@ -44,10 +44,12 @@ CREDIT_LINE("ODBC", "Stig Bakken, Andreas Karajannis, Frank M. Kromann, Daniel R
 CREDIT_LINE("Opcache", "Andi Gutmans, Zeev Suraski, Stanislav Malyshev, Dmitry Stogov, Xinchen Hui");
 CREDIT_LINE("OpenSSL", "Stig Venaas, Wez Furlong, Sascha Kettler, Scott MacVicar, Eliot Lear");
 CREDIT_LINE("pcntl", "Jason Greene, Arnaud Le Blanc");
+CREDIT_LINE("pdo-pglite", "Sean Morris");
 CREDIT_LINE("Perl Compatible Regexps", "Andrei Zmievski");
 CREDIT_LINE("PHP Archive", "Gregory Beaver, Marcus Boerger");
 CREDIT_LINE("PHP Data Objects", "Wez Furlong, Marcus Boerger, Sterling Hughes, George Schlossnagle, Ilia Alshanetsky");
 CREDIT_LINE("PHP hash", "Sara Golemon, Rasmus Lerdorf, Stefan Esser, Michael Wallner, Scott MacVicar");
+CREDIT_LINE("pib", "Oraoto, Sean Morris");
 CREDIT_LINE("Posix", "Kristian Koehntopp");
 CREDIT_LINE("PostgreSQL driver for PDO", "Edin Kadribasic, Ilia Alshanetsky");
 CREDIT_LINE("PostgreSQL", "Jouni Ahto, Zeev Suraski, Yasuo Ohgaki, Chris Kings-Lynne");
@@ -69,6 +71,7 @@ CREDIT_LINE("System V Semaphores", "Tom May");
 CREDIT_LINE("System V Shared Memory", "Christian Cartus");
 CREDIT_LINE("tidy", "John Coggeshall, Ilia Alshanetsky");
 CREDIT_LINE("tokenizer", "Andrei Zmievski, Johannes Schlueter");
+CREDIT_LINE("Vrzno", "Sean Morris");
 CREDIT_LINE("XML", "Stig Bakken, Thies C. Arntzen, Sterling Hughes");
 CREDIT_LINE("XMLReader", "Rob Richards");
 CREDIT_LINE("XMLWriter", "Rob Richards, Pierre-Alain Joye");
diff --git a/third_party/php8.4-src/ext/tokenizer/tokenizer.c b/third_party/php8.4-src/ext/tokenizer/tokenizer.c
index 9d1b19d..94c3f5a 100644
--- a/third_party/php8.4-src/ext/tokenizer/tokenizer.c
+++ b/third_party/php8.4-src/ext/tokenizer/tokenizer.c
@@ -318,7 +318,7 @@ static void add_token(
 	zend_hash_next_index_insert_new(Z_ARRVAL_P(return_value), &token);
 }
 
-static bool tokenize(zval *return_value, zend_string *source, zend_class_entry *token_class)
+bool tokenize(zval *return_value, zend_string *source, zend_class_entry *token_class)
 {
 	zval source_zval;
 	zend_lex_state original_lex_state;
diff --git a/third_party/php8.4-src/main/fastcgi.c b/third_party/php8.4-src/main/fastcgi.c
index 448576a..7209487 100644
--- a/third_party/php8.4-src/main/fastcgi.c
+++ b/third_party/php8.4-src/main/fastcgi.c
@@ -551,6 +551,7 @@ void fcgi_shutdown(void)
 {
 	if (is_initialized) {
 		zend_hash_destroy(&fcgi_mgmt_vars);
+		is_initialized = 0;
 	}
 	is_fastcgi = 0;
 	if (allowed_clients) {
diff --git a/third_party/php8.4-src/sapi/cgi/Makefile.frag b/third_party/php8.4-src/sapi/cgi/Makefile.frag
index 79e2afe..0956f43 100644
--- a/third_party/php8.4-src/sapi/cgi/Makefile.frag
+++ b/third_party/php8.4-src/sapi/cgi/Makefile.frag
@@ -1,7 +1,7 @@
 cgi: $(SAPI_CGI_PATH)
 
 $(SAPI_CGI_PATH): $(PHP_GLOBAL_OBJS) $(PHP_BINARY_OBJS) $(PHP_FASTCGI_OBJS) $(PHP_CGI_OBJS)
-	$(BUILD_CGI)
+	$(BUILD_CGI) $(WASM_SHARED_LIBS)
 
 install-cgi: $(SAPI_CGI_PATH)
 	@echo "Installing PHP CGI binary:        $(INSTALL_ROOT)$(bindir)/"
diff --git a/third_party/php8.4-src/sapi/cgi/cgi_main.c b/third_party/php8.4-src/sapi/cgi/cgi_main.c
index 0e7b300..3652b52 100644
--- a/third_party/php8.4-src/sapi/cgi/cgi_main.c
+++ b/third_party/php8.4-src/sapi/cgi/cgi_main.c
@@ -18,7 +18,7 @@
    |          Dmitry Stogov <dmitry@php.net>                              |
    +----------------------------------------------------------------------+
 */
-
+#include <emscripten.h>
 #include "php.h"
 #include "php_globals.h"
 #include "php_variables.h"
@@ -1872,7 +1872,8 @@ int main(int argc, char *argv[])
 	SG(request_info).path_translated = NULL;
 #endif
 
-	cgi_sapi_module.executable_location = argv[0];
+	// cgi_sapi_module.executable_location = argv[0];
+	cgi_sapi_module.executable_location = NULL;
 	if (!cgi && !fastcgi && !bindpath) {
 		cgi_sapi_module.additional_functions = additional_functions;
 	}
@@ -1954,8 +1955,10 @@ consult the installation file that came with this distribution, or visit \n\
 	}
 
 	/* make php call us to get _ENV vars */
-	php_php_import_environment_variables = php_import_environment_variables;
-	php_import_environment_variables = cgi_php_import_environment_variables;
+	if(php_import_environment_variables != cgi_php_import_environment_variables) {
+		php_php_import_environment_variables = php_import_environment_variables;
+		php_import_environment_variables = cgi_php_import_environment_variables;
+	}
 
 	if (fastcgi) {
 		/* How many times to run PHP scripts before dying */
@@ -2681,3 +2684,18 @@ consult the installation file that came with this distribution, or visit \n\
 	return exit_status;
 }
 /* }}} */
+
+void EMSCRIPTEN_KEEPALIVE wasm_sapi_cgi_init(void)
+{
+	putenv("USE_ZEND_ALLOC=0");
+}
+
+char* EMSCRIPTEN_KEEPALIVE wasm_sapi_cgi_getenv(char *name)
+{
+	return getenv(name);
+}
+
+char* EMSCRIPTEN_KEEPALIVE wasm_sapi_cgi_putenv(char *name, char *value)
+{
+	return _sapi_cgi_putenv(name, strlen(name), value);
+}
diff --git a/third_party/php8.4-src/sapi/cli/Makefile.frag b/third_party/php8.4-src/sapi/cli/Makefile.frag
index aa1d642..649533d 100644
--- a/third_party/php8.4-src/sapi/cli/Makefile.frag
+++ b/third_party/php8.4-src/sapi/cli/Makefile.frag
@@ -1,7 +1,7 @@
 cli: $(SAPI_CLI_PATH)
 
 $(SAPI_CLI_PATH): $(PHP_GLOBAL_OBJS) $(PHP_BINARY_OBJS) $(PHP_CLI_OBJS)
-	$(BUILD_CLI)
+	$(BUILD_CLI) $(WASM_SHARED_LIBS)
 
 install-cli: $(SAPI_CLI_PATH)
 	@echo "Installing PHP CLI binary:        $(INSTALL_ROOT)$(bindir)/"
diff --git a/third_party/php8.4-src/sapi/embed/php_embed.c b/third_party/php8.4-src/sapi/embed/php_embed.c
index c18480d..787687f 100644
--- a/third_party/php8.4-src/sapi/embed/php_embed.c
+++ b/third_party/php8.4-src/sapi/embed/php_embed.c
@@ -242,8 +242,8 @@ EMBED_SAPI_API int php_embed_init(int argc, char **argv)
 		return FAILURE;
 	}
 
-	SG(headers_sent) = 1;
-	SG(request_info).no_headers = 1;
+	SG(headers_sent) = 0;
+	SG(request_info).no_headers = 0;
 	php_register_variable("PHP_SELF", "-", NULL);
 
 	return SUCCESS;
diff --git a/third_party/php8.4-src/sapi/phpdbg/Makefile.frag b/third_party/php8.4-src/sapi/phpdbg/Makefile.frag
index a069a236..e37658af 100644
--- a/third_party/php8.4-src/sapi/phpdbg/Makefile.frag
+++ b/third_party/php8.4-src/sapi/phpdbg/Makefile.frag
@@ -6,7 +6,7 @@ $(BUILD_SHARED): $(PHP_GLOBAL_OBJS) $(PHP_BINARY_OBJS) $(PHP_PHPDBG_OBJS)
 	$(BUILD_PHPDBG_SHARED)
 
 $(BUILD_BINARY): $(PHP_GLOBAL_OBJS) $(PHP_BINARY_OBJS) $(PHP_PHPDBG_OBJS)
-	$(BUILD_PHPDBG)
+	$(BUILD_PHPDBG) $(WASM_SHARED_LIBS)
 
 %.c: %.y
 %.c: %.l
diff --git a/third_party/php8.4-src/sapi/phpdbg/phpdbg.c b/third_party/php8.4-src/sapi/phpdbg/phpdbg.c
index 76b3c7f..b8c80bd 100644
--- a/third_party/php8.4-src/sapi/phpdbg/phpdbg.c
+++ b/third_party/php8.4-src/sapi/phpdbg/phpdbg.c
@@ -16,6 +16,8 @@
    +----------------------------------------------------------------------+
 */
 
+#include  <emscripten.h>
+
 #include "phpdbg.h"
 #include "phpdbg_prompt.h"
 #include "phpdbg_bp.h"
@@ -149,6 +151,24 @@ static inline void php_phpdbg_globals_ctor(zend_phpdbg_globals *pg) /* {{{ */
 #endif
 } /* }}} */
 
+const char* phpdbg_wasm_current_file = NULL;
+uint32_t phpdbg_wasm_current_line = NULL;
+
+const char** EMSCRIPTEN_KEEPALIVE phpdbg_wasm_get_current_file(void)
+{
+	return &phpdbg_wasm_current_file;
+}
+
+uint32_t* EMSCRIPTEN_KEEPALIVE phpdbg_wasm_get_current_line(void)
+{
+	return &phpdbg_wasm_current_line;
+}
+
+int* EMSCRIPTEN_KEEPALIVE php_wasm_get_bpcount(void)
+{
+	return &PHPDBG_G(bp_count);
+}
+
 static PHP_MINIT_FUNCTION(phpdbg) /* {{{ */
 {
 	zend_hash_init(&PHPDBG_G(bp)[PHPDBG_BREAK_FILE], 8, NULL, php_phpdbg_destroy_bp_file, 0);
@@ -178,6 +198,9 @@ static PHP_MINIT_FUNCTION(phpdbg) /* {{{ */
 
 static PHP_MSHUTDOWN_FUNCTION(phpdbg) /* {{{ */
 {
+	phpdbg_wasm_current_file = NULL;
+	phpdbg_wasm_current_line = NULL;
+
 	zend_hash_destroy(&PHPDBG_G(registered));
 	phpdbg_destroy_watchpoints();
 
@@ -1119,7 +1142,7 @@ php_stream *phpdbg_stream_url_wrap_php(php_stream_wrapper *wrapper, const char *
 	return PHPDBG_G(orig_url_wrap_php)->wops->stream_opener(wrapper, path, mode, options, opened_path, context STREAMS_CC);
 } /* }}} */
 
-int main(int argc, char **argv) /* {{{ */
+int EMSCRIPTEN_KEEPALIVE main(int argc, char **argv) /* {{{ */
 {
 	sapi_module_struct *phpdbg = &phpdbg_sapi_module;
 	char *sapi_name;
diff --git a/third_party/php8.4-src/sapi/phpdbg/phpdbg.h b/third_party/php8.4-src/sapi/phpdbg/phpdbg.h
index 41fe8fa..b4fea58 100644
--- a/third_party/php8.4-src/sapi/phpdbg/phpdbg.h
+++ b/third_party/php8.4-src/sapi/phpdbg/phpdbg.h
@@ -182,7 +182,7 @@ int phpdbg_do_parse(phpdbg_param_t *stack, char *input);
 #ifndef _WIN32
 #	define PHPDBG_DEFAULT_FLAGS (PHPDBG_IS_QUIET | PHPDBG_IS_COLOURED | PHPDBG_IS_BP_ENABLED | PHPDBG_HAS_PAGINATION)
 #else
-#	define PHPDBG_DEFAULT_FLAGS (PHPDBG_IS_QUIET | PHPDBG_IS_BP_ENABLED | PHPDBG_HAS_PAGINATION)
+#	define PHPDBG_DEFAULT_FLAGS (PHPDBG_IS_QUIET | PHPDBG_IS_BP_ENABLED)
 #endif /* }}} */
 
 /* {{{ output descriptors */
@@ -311,4 +311,7 @@ ZEND_BEGIN_MODULE_GLOBALS(phpdbg)
 	zend_ulong lines;                                  /* max number of lines to display */
 ZEND_END_MODULE_GLOBALS(phpdbg) /* }}} */
 
+extern const char* phpdbg_wasm_current_file;
+extern uint32_t phpdbg_wasm_current_line;
+
 #endif /* PHPDBG_H */
diff --git a/third_party/php8.4-src/sapi/phpdbg/phpdbg_cmd.c b/third_party/php8.4-src/sapi/phpdbg/phpdbg_cmd.c
index 9877d60..fb181d2 100644
--- a/third_party/php8.4-src/sapi/phpdbg/phpdbg_cmd.c
+++ b/third_party/php8.4-src/sapi/phpdbg/phpdbg_cmd.c
@@ -774,7 +774,7 @@ PHPDBG_API char *phpdbg_read_input(const char *buffered) /* {{{ */
 			}
 #else
 			char buf[PHPDBG_MAX_CMD];
-			phpdbg_write("%s", phpdbg_get_prompt());
+			// phpdbg_write("%s", phpdbg_get_prompt());
 			phpdbg_consume_stdin_line(buf);
 			buffer = estrdup(buf);
 #endif
diff --git a/third_party/php8.4-src/sapi/phpdbg/phpdbg_io.c b/third_party/php8.4-src/sapi/phpdbg/phpdbg_io.c
index 261a67c..d614628 100644
--- a/third_party/php8.4-src/sapi/phpdbg/phpdbg_io.c
+++ b/third_party/php8.4-src/sapi/phpdbg/phpdbg_io.c
@@ -14,6 +14,8 @@
    +----------------------------------------------------------------------+
 */
 
+#include  <emscripten.h>
+
 #ifdef HAVE_CONFIG_H
 #include <config.h>
 #endif
@@ -22,52 +24,40 @@
 
 ZEND_EXTERN_MODULE_GLOBALS(phpdbg)
 
-/* is easy to generalize ... but not needed for now */
-PHPDBG_API int phpdbg_consume_stdin_line(char *buf) {
-	int bytes = PHPDBG_G(input_buflen), len = 0;
+EM_ASYNC_JS(int, phpdbg_internal_consume_stdin_line, (char *buf, int max_length), {
+
+	let input;
+
+	Module.triggerStdin();
 
-	if (PHPDBG_G(input_buflen)) {
-		memcpy(buf, PHPDBG_G(input_buffer), bytes);
+	if(Module.inputDataQueue.length)
+	{
+		input = String(Module.inputDataQueue.shift());
+	}
+	else
+	{
+		let a, r;
+		const promise = new Promise((accept, reject) => [a, r] = [ accept, reject ]);
+		Module.awaitingInput = a;
+		input = String(await promise);
 	}
 
-	PHPDBG_G(last_was_newline) = 1;
+	if(input.length >= max_length)
+	{
+		input = input.slice(0, max_length - 1);
+	}
 
-	do {
-		int i;
-		if (bytes <= 0) {
-			continue;
-		}
+	stringToUTF8(input, buf, input.length + 1);
 
-		for (i = len; i < len + bytes; i++) {
-			if (buf[i] == '\x03') {
-				if (i != len + bytes - 1) {
-					memmove(buf + i, buf + i + 1, len + bytes - i - 1);
-				}
-				len--;
-				i--;
-				continue;
-			}
-			if (buf[i] == '\n') {
-				PHPDBG_G(input_buflen) = len + bytes - 1 - i;
-				if (PHPDBG_G(input_buflen)) {
-					memcpy(PHPDBG_G(input_buffer), buf + i + 1, PHPDBG_G(input_buflen));
-				}
-				if (i != PHPDBG_MAX_CMD - 1) {
-					buf[i + 1] = 0;
-				}
-				return i;
-			}
-		}
 
-		len += bytes;
-	} while ((bytes = phpdbg_mixed_read(PHPDBG_G(io)[PHPDBG_STDIN].fd, buf + len, PHPDBG_MAX_CMD - len, -1)) > 0);
 
-	if (bytes <= 0) {
-		PHPDBG_G(flags) |= PHPDBG_IS_QUITTING;
-		zend_bailout();
-	}
+	return input.length;
+});
 
-	return bytes;
+
+/* is easy to generalize ... but not needed for now */
+PHPDBG_API int phpdbg_consume_stdin_line(char *buf) {
+	return phpdbg_internal_consume_stdin_line(buf, PHPDBG_MAX_CMD);
 }
 
 PHPDBG_API int phpdbg_mixed_read(int fd, char *ptr, int len, int tmo) {
diff --git a/third_party/php8.4-src/sapi/phpdbg/phpdbg_prompt.c b/third_party/php8.4-src/sapi/phpdbg/phpdbg_prompt.c
index 6333fdc..d709eb3 100644
--- a/third_party/php8.4-src/sapi/phpdbg/phpdbg_prompt.c
+++ b/third_party/php8.4-src/sapi/phpdbg/phpdbg_prompt.c
@@ -1662,6 +1662,8 @@ void phpdbg_execute_ex(zend_execute_data *execute_data) /* {{{ */
 #endif /* ZEND_CHECK_STACK_LIMIT */
 
 	while (1) {
+		phpdbg_wasm_current_file = zend_get_executed_filename();
+		phpdbg_wasm_current_line = zend_get_executed_lineno();
 		zend_object *exception = EG(exception);
 
 		if ((PHPDBG_G(flags) & PHPDBG_BP_RESOLVE_MASK)) {
diff --git a/third_party/php8.4-src/sapi/phpdbg/phpdbg_utils.c b/third_party/php8.4-src/sapi/phpdbg/phpdbg_utils.c
index 329ee9a..2b6d83e 100644
--- a/third_party/php8.4-src/sapi/phpdbg/phpdbg_utils.c
+++ b/third_party/php8.4-src/sapi/phpdbg/phpdbg_utils.c
@@ -16,6 +16,8 @@
    +----------------------------------------------------------------------+
 */
 
+#include <emscripten.h>
+
 #include "zend.h"
 
 #include "php.h"
@@ -293,7 +295,7 @@ PHPDBG_API void phpdbg_set_prompt(const char *prompt) /* {{{ */
 	PHPDBG_G(prompt)[0] = strdup(prompt);
 } /* }}} */
 
-PHPDBG_API const char *phpdbg_get_prompt(void) /* {{{ */
+PHPDBG_API const char EMSCRIPTEN_KEEPALIVE *phpdbg_get_prompt(void) /* {{{ */
 {
 	/* find cached prompt */
 	if (PHPDBG_G(prompt)[1]) {
